packages/connector-azuread/src/index.ts

```diff
@@ -0,0 +1,177 @@
+import {
+  ConfidentialClientApplication,
+  AuthorizationCodeRequest,
+  AuthorizationUrlRequest,
+  CryptoProvider,
+} from '@azure/msal-node';
+import {
+  ConnectorError,
+  ConnectorErrorCodes,
+  GetAuthorizationUri,
+  GetUserInfo,
+  ConnectorMetadata,
+  Connector,
+  SocialConnectorInstance,
+  GetConnectorConfig,
+  codeWithRedirectDataGuard,
+} from '@logto/connector-types';
+import { assert } from '@silverhand/essentials';
+import got, { HTTPError } from 'got';
+
+import { scopes, defaultMetadata, defaultTimeout, graphAPIEndpoint } from './constant';
+import {
+  azureADConfigGuard,
+  AzureADConfig,
+  accessTokenResponseGuard,
+  userInfoResponseGuard,
+} from './types';
+
+export default class AzureADConnector implements SocialConnectorInstance<AzureADConfig> {
+  public metadata: ConnectorMetadata = defaultMetadata;
+
+  public clientApplication!: ConfidentialClientApplication;
+  public authCodeUrlParams!: AuthorizationUrlRequest;
+  public verifier!: string;
+  public challenge!: string;
+
+  cryptoProvider = new CryptoProvider();
+  private readonly authCodeRequest!: AuthorizationCodeRequest;
+
+  private _connector?: Connector;
+
+  public get connector() {
+    if (!this._connector) {
+      throw new ConnectorError(ConnectorErrorCodes.General);
+    }
+
+    return this._connector;
+  }
+
+  public set connector(input: Connector) {
+    this._connector = input;
+  }
+
+  constructor(public readonly getConfig: GetConnectorConfig) {}
+
+  public validateConfig(config: unknown): asserts config is AzureADConfig {
+    const result = azureADConfigGuard.safeParse(config);
+
+    if (!result.success) {
+      throw new ConnectorError(ConnectorErrorCodes.InvalidConfig, result.error);
+    }
+  }
+
+  public getAuthorizationUri: GetAuthorizationUri = async ({ state, redirectUri }) => {
+    const config = await this.getConfig(this.metadata.id);
+
+    this.validateConfig(config);
+    const { clientId, clientSecret, cloudInstance, tennantId } = config;
+
+    this.authCodeUrlParams = {
+      scopes,
+      state,
+      redirectUri,
+    };
+
+    this.clientApplication = new ConfidentialClientApplication({
+      auth: {
+        clientId,
+        clientSecret,
+        authority: cloudInstance + tennantId,
+      },
+    });
+
+    const { challenge, verifier } = await this.cryptoProvider.generatePkceCodes();
+    this.challenge = challenge;
+    this.verifier = verifier;
```

Comment:
there will be only one connector instance per Logto instance, so if we have to use PKCE flow, we need to store the verifier in a centralized cache or frontend, that connects with the session instead of using a class property. otherwise all users will share the same `verifier`, which is not what we want.

also imagine you have multiple Logto instances, we may cannot find the `verifier` if the `.getUserInfo()` request hits a different instance.

@simeng-li @wangsijie any ideas?

Reply:
Yes, I believe PKCE should be a session scoped variable, the code_verifier and challenge should be unique per authorization request.  Also, feels like it should be generated/kept at the client side only. Hmm, open for discussion.
 
https://datatracker.ietf.org/doc/html/rfc7636

Reply:
Hey, BTW, @ufuf just checked the msal-node docs. 

Since we are generating a [ConfidentialClientApplication](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/docs/initialize-confidential-client-application.md) here,  we should be able to use client credentials to complete the auth-code exchange flow. 
https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/auth-code

PKCE is used for [public applications](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/docs/initialize-public-client-application.md) without client credentials. 
https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/auth-code-pkce

We could always treat Logto as a pre-registered ConfidentialClientApplication under this circumstance?  Let me know if this makes sense to you.

Reply:
You are right, we do not need PKCE for confidential applications.
I removed and tested it 👍